### ISSUE TITLE
How to solve the blurred warped image issue?

### ISSUE THREAD

**Issue by shuo-shuo-zhang** [voxelmorph/voxelmorph#43]
Dear Dr. Dalca,
I am so sorry I haven’t solved my problem about the blurred warped images. I use Colin27 volumes as my atlas and ADNI volumes as the subject images. The maximum intensity of the Colin27 volumes is about 0.62 and that of the ADNI volumes is about 0.86 after intensity normalization to [0,1] by simply using “ img = img/255 ” and the minimum intensity for both is 0. 

The model I am using is your miccai voxelmorph, and the prior_lambda is 15, the image_sigma is 0.02, batch size is 1, the rest of parameters are as default. I find that no matter how I change the two sigma and lambda parameters, the loss values are no less than 8 during training process. I use the trained model to generate deformation field, and then use the deformation field to warp the Colin volume, I find that the warped images are all blurred, I don’t know the specific reason for that. I really appreciate that if you can help me with that, thank you very much.

Best wishes,
-Shuo

**Comment by adalca** [comment]
@shuo-shuo-zhang ,

The Colin27 and ADNI data seem to be on different intensity scales (judging by the maximums you provided). For the MICCAI model, you could try to intensity match them, but the easier thing to do would be to try training the CVPR version with CC loss function, which will be robust to intensity variations like this. Can you try this and let us know how it goes?

**Comment by shuo-shuo-zhang** [comment]
Dear Dr. Dalca,
Yes, I will try this and let you know it as soon as I get the result. But I want to know if I can still try the MICCAI version and use the NCC loss function. Do you have any suggestions on this? Thank you very much.

**Comment by adalca** [comment]
yes, you can. You can simply replace the MSE part of the term (`loss_class.recon_loss`) with an NCC.  The `prior_lambda` would have to be adjusted.

are you interested in the MICCAI term because of the diffeomorphic aspect, or the probabilistic? (or both?)

**Comment by shuo-shuo-zhang** [comment]
Actually, I am interested in both of them, I think your work is incredible and I hope I can get a good model which can be diffeomorphic and robust. Thank you very much. I will let you know if I have a good result.

**Comment by adalca** [comment]
The diffeomorphic aspect can be done without the miccai version  -- you can simply add the velocity field integration to the CVPR model. We're hoping to do this and put the code up there in the next month or so. Until then, you can try the modification mentioned above.

**Comment by shuo-shuo-zhang** [comment]
Thanks, I will try the modification as soon as possible. May I ask if I don't need the probabilistic aspect, then do I not need to generate the velocity field variance(flow_log_sigma)? That is, all I need to do is to get the velocity field mean by the u-net model and then use the velocity field integration to directly integrate the velocity field mean? Does this can guarantee the diffeomorphic property? Thank you very much for all your help.

**Comment by adalca** [comment]
yup, exactly!

In our work, we integrate (with `VecInt` layer) the field at half-resolution to save on memory issues. 

The diffeomorphic guarantee is in the continuous domain, while in practice it depends on numerical aspects, such as the extent of the field and the number of integration steps (e.g. if you use ss), or ode solver parameters (if you use `odeint`). In our experiments (see section 3.4.2 in https://arxiv.org/pdf/1903.03545.pdf), after a few scaling and squaring steps, the fields are diffeomorphic nearly everywhere.

**Comment by shuo-shuo-zhang** [comment]
Thank you very much for your reply! I think I know what I should do next, if I get something new in later time, I will let you know.

**Comment by shuo-shuo-zhang** [comment]
Dear Dr. Dalca,
I have tried the miccai version with NCC loss function, after more than 300 epochs, the warped images are better than before, but still blurred. But in the process of test, the dice score for all brain structures of all subjects is lower. That means the registration result is not as accurate as before when using the mse loss function, even the blurred images have been improved a little. I don't know if it's worth to improve the warped image quality by sacrificing the registration accuracy.

May I ask if I can try some other means to ensure both the registration accuracy and the warped image quality? Last time you told me to try to  intensity match the image pairs. Does "intensity match" mean  histogram matching or something? Can you show me some methods to do this simply? Thank you very much.
- Shuo

**Comment by shuo-shuo-zhang** [comment]
Dear Dr. Dalca,
I am so sorry,  @adalca , I haven't solved my problem, I want to know if you can recommend me some suggestions about how to perform histogram matching, I have tried some histogram matching methods, but the quality of warped images is worse than before, I don't know what to do. Is histogram matching not good for the blurred images? 

I tried to see the deformation field, it's extremely unsmooth. Is that because the image sigma is not small enough? So I should use some smaller image sigma values. Thank you for all your help.

**Comment by adalca** [comment]
Can you please show me example images and deformations for the NCC version that you ran above? If the deformation field is unsmooth, you should change the velocity field parameter to encourage more smoothness (depending on which version you are using this parameter should likely go up -- which loss are you using?)
